cmd/manager/main.go

```diff
@@ -19,78 +19,71 @@ package main
-	"sigs.k8s.io/cluster-api-provider-aws/pkg/record"
-	capi "sigs.k8s.io/cluster-api/pkg/apis"
-	"sigs.k8s.io/controller-runtime/pkg/client/config"
-	"sigs.k8s.io/controller-runtime/pkg/manager"
-	"sigs.k8s.io/controller-runtime/pkg/runtime/log"
-	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/apis/infrastructure/v1alpha2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	// +kubebuilder:scaffold:imports
 )
 
+var (
+	myscheme = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+func init() {
+	_ = scheme.AddToScheme(myscheme)
+	_ = v1alpha2.AddToScheme(myscheme)
+	// +kubebuilder:scaffold:scheme
+}
+
 func main() {
-	klog.InitFlags(nil)
-	flag.Set("logtostderr", "true")
+	var metricsAddr string
+	var enableLeaderElection bool
 
+	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
+	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
+		"Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.")
 	watchNamespace := flag.String("namespace", "",
 		"Namespace that the controller watches to reconcile cluster-api objects. If unspecified, the controller watches for cluster-api objects across all namespaces.")
-
 	profilerAddress := flag.String("profiler-address", "", "Bind address to expose the pprof profiler (e.g. localhost:6060)")
-
-	metricsAddress := flag.String("metrics-address", "", "Bind address to expose the metrics (e.g. localhost:8080)")
-
 	flag.Parse()
+
 	if *watchNamespace != "" {
-		klog.Infof("Watching cluster-api objects only in namespace %q for reconciliation", *watchNamespace)
+		setupLog.Info("Watching cluster-api objects only in namespace for reconciliation", "namespace", *watchNamespace)
 	}
 
 	if *profilerAddress != "" {
-		klog.Infof("Profiler listening for requests at %s", *profilerAddress)
+		setupLog.Info("Profiler listening for requests", "profiler-address", *profilerAddress)
 		go func() {
-			klog.Info(http.ListenAndServe(*profilerAddress, nil))
+			setupLog.Error(http.ListenAndServe(*profilerAddress, nil), "listen and serve error")
 		}()
 	}
 
-	// Setup a Manager
 	syncPeriod := 10 * time.Minute
 
-	// Setup controller-runtime logger.
-	log.SetLogger(klogr.New())
+	ctrl.SetLogger(klogr.New())
 
-	// Get a config to talk to the api-server.
-	cfg := config.GetConfigOrDie()
-	mgr, err := manager.New(cfg, manager.Options{
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:             myscheme,
+		MetricsBindAddress: metricsAddr,
+		LeaderElection:     enableLeaderElection,
 		SyncPeriod:         &syncPeriod,
-		Namespace:          *watchNamespace,
-		MetricsBindAddress: *metricsAddress,
 	})
 	if err != nil {
-		klog.Fatalf("Failed to set up overall controller manager: %v", err)
-	}
-
-	// Initialize event recorder.
-	record.InitFromRecorder(mgr.GetEventRecorderFor("aws-controller"))
```

Comment:
also still needed

Reply:
Should it, or is it something that we should set in the individual controllers instead?

Reply:
oh i'm not sure. will it still work initialized in main? I'd like to keep it there and then move it later